### PR TITLE
Decoded AnimationControllerComponent properties and related structs

### PR DESCRIPTION
**Pull request by Mozz3d**

### CODE REVIEW

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+// clang-format off
+
+// This file is generated from the Game's Reflection data
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:5]: Please remove comments that disable format checking and say the file is generated :)

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+struct AnimationControllerComponent;
+
+struct IKTargetController
+{
+    struct IKTargetData
+    {
+        anim::IKTargetRef ikTargetref;                    // 00
+        Handle<IPositionProvider> positionProvider;       // 10
+        Handle<IOrientationProvider> orientationProvider; // 20
+    };
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:35]: Inner structures should also be checked with `RED4EXT_ASSERT_SIZE`.

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        anim::LookAtRef lookAtRef;                  // 00
+        Handle<IPositionProvider> positionProvider; // 10
+        uint32_t unk20;                             // 20
+    };
+    struct AdditionalLookatdata
+    {
+        CName partName;                       // 00                    
+        DynArray<anim::LookAtRef> lookAtRefs; // 08
+    };
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:55]: Inner structures should also be checked with `RED4EXT_ASSERT_SIZE`.

**Review by psiberx — CHANGES_REQUESTED** [review]

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+struct IKTargetController
+{
+    struct IKTargetData
+    {
+        anim::IKTargetRef ikTargetref;                    // 00
+        Handle<IPositionProvider> positionProvider;       // 10
+        Handle<IOrientationProvider> orientationProvider; // 20
+    };
+    AnimationControllerComponent& animationControllerComponent; // 00
+    uint64_t unk08;                                             // 08
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:37]: Maybe it's `Handle<AnimationControllerComponent>`? Same for `LookAtController`.

Comment (Mozz3d) [reply include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:37]: It's just a pointer in memory, so I left it as a reference. Not sure Handle would fit here

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+namespace ent
+{
+struct AnimationControlBinding;
+struct AnimationControllerComponent;
+
+struct IKTargetController
+{
+    struct IKTargetData
+    {
+        anim::IKTargetRef ikTargetref;                    // 00
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:31]: Naming: `ikTargetRef`

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        Handle<IPositionProvider> positionProvider;       // 10
+        Handle<IOrientationProvider> orientationProvider; // 20
+    };
+    RED4EXT_ASSERT_SIZE(IKTargetData, 0x30);
+
+    AnimationControllerComponent& animationControllerComponent; // 00
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:37]: It seems more practical to make it `AnimationControllerComponent*`, at least until there is a code that initializes or manages the instance.

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    uint64_t unk08;                                             // 08
+    DynArray<IKTargetData> IKTargetData;                        // 10
+    Handle<anim::IKTargetParams_Update> IKparams;               // 20
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:40]: Naming: `ikTargetData` and `ikParams`

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+struct LookAtController
+{
+    struct LookatData
+    {
+        anim::LookAtRef lookAtRef;                  // 00
+        Handle<IPositionProvider> positionProvider; // 10
+        uint32_t unk20;                             // 20
+    };
+    RED4EXT_ASSERT_SIZE(LookatData, 0x28);
+    struct AdditionalLookatData
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:53]: Naming: `AdditionalLookAtData`
There should be empty line between structs.

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    AnimationControllerComponent& animationControllerComponent; // 00
+    uint64_t unk08;                                             // 08
+    DynArray<IKTargetData> IKTargetData;                        // 10
+    Handle<anim::IKTargetParams_Update> IKparams;               // 20
+};
+RED4EXT_ASSERT_SIZE(IKTargetController, 0x30);
+
+struct LookAtController
+{
+    struct LookatData
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:46]: Naming: `LookAtData`

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    };
+    RED4EXT_ASSERT_SIZE(LookatData, 0x28);
+    struct AdditionalLookatData
+    {
+        CName partName;                       // 00
+        DynArray<anim::LookAtRef> lookAtRefs; // 08
+    };
+    RED4EXT_ASSERT_SIZE(AdditionalLookatData, 0x28);
+
+    AnimationControllerComponent& animationControllerComponent; // 00
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:60]: Same case as with `IKTargetController`.

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    uint32_t unkCounter;                                        // 08
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:61]: Naming: `unk08` or `counter`
You can add a comment if you know something about the counter.

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    void* unk10;                                                // 10
+    bool usingTPPCamera;                                        // 18
+    DynArray<LookatData> lookAtData;                            // 20
+    DynArray<AdditionalLookatData> additionalLookatdata;        // 30
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:65]: Naming: `additionalLookAtData`

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    Handle<anim::LookAtParams_UpdatePositions> lookAtparams;    // 40
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:66]: Naming: `lookAtParams`

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+struct AnimationControllerComponent : ent::IComponent
+{
+    static constexpr const char* NAME = "entAnimationControllerComponent";
+    static constexpr const char* ALIAS = "AnimationControllerComponent";
+
+    uint64_t unk90[7];                                   // 90
+    void* unkC8;                                         // C8
+    void* unkD0;                                         // D0
+    DynArray<uint64_t> unkD8Array;                       // D8
+    DynArray<uint64_t> unkE8Array;                       // E8
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:83]: Naming: `unkD8` and `unkE8`

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+    Handle<ent::AnimationControlBinding> controlBinding; // 1E0
+};
+RED4EXT_ASSERT_SIZE(AnimationControllerComponent, 0x1F0);
+
+} // namespace ent
+using entAnimationControllerComponent = ent::AnimationControllerComponent;
+using AnimationControllerComponent = ent::AnimationControllerComponent;
+} // namespace RED4ext
+
+// clang-format on
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:98]: Delete this line.

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+    uint64_t unk90[7];                                   // 90
+    void* unkC8;                                         // C8
+    void* unkD0;                                         // D0
+    DynArray<uint64_t> unkD8Array;                       // D8
+    DynArray<uint64_t> unkE8Array;                       // E8
+    Ref<anim::ActionAnimDatabase> actionAnimDatabaseRef; // F8
+    anim::AnimDatabaseCollection animDatabaseCollection; // 110
+    LookAtController lookAtController;                   // 120
+    IKTargetController ikTargetController;               // 1A0
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:87]: If `LookAtController` is `0x88` bytes, then `0x120 + 0x88 = 0x1A8`.
Please add `RED4EXT_ASSERT_OFFSET` for all named fields of `AnimationControllerComponent` and some fields of other structs.

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    uint64_t unk80;                                             // 80
+};
+RED4EXT_ASSERT_SIZE(LookAtController, 0x88);
+
+struct AnimationControllerComponent : ent::IComponent
+{
+    static constexpr const char* NAME = "entAnimationControllerComponent";
+    static constexpr const char* ALIAS = "AnimationControllerComponent";
+
+    uint64_t unk90[7];                                   // 90
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:79]: If this area is not yet decoded, then mark it as `uint8_t unk90[0xC8 - 0x90];`.

File: include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+    RED4EXT_ASSERT_SIZE(AdditionalLookatData, 0x28);
+
+    AnimationControllerComponent& animationControllerComponent; // 00
+    uint32_t unkCounter;                                        // 08
+    void* unk10;                                                // 10
+    bool usingTPPCamera;                                        // 18
+    DynArray<LookatData> lookAtData;                            // 20
+    DynArray<AdditionalLookatData> additionalLookatdata;        // 30
+    Handle<anim::LookAtParams_UpdatePositions> lookAtparams;    // 40
+    uint64_t unk50[5];                                          // 50
```
Comment (psiberx) [inline include/RED4ext/Scripting/Natives/entAnimationControllerComponent.hpp:67]: If this area is not yet decoded, then mark it as `uint8_t unk50[0x78 - 0x50];`.

**Review by psiberx — CHANGES_REQUESTED** [review]

**Review by psiberx — CHANGES_REQUESTED** [review]

**Review by psiberx — APPROVED** [review]

**Review by WopsS — APPROVED** [review]

**Pull request merged**